### PR TITLE
Script engine merge 89

### PR DESCRIPTION
**Pull request by cesare-montresor**
- Scripts
  - now handles only the UI ( perhaps something is left behind, but it shouldn't be )
  
- ScriptsEnhanced
  - moved away from Scripts
  - ScriptsEnhanced  handles all instanced of ScriptsEnhanced and their threads
  
- ScriptsEnhancedEngine
  - takes care of selecting, initializing, preloading and executing a scripts using the correct engine
  
- ScriptsRecorder
  - no longer static/singleton
  
- ScriptsRecorderService
  - hanadle instances
  - collect and broadcast messages to Recorders
  
- ScriptsEnhancedEditor:
  - now use regular ScriptsEnhanced rather then custom thread/PythonEngine
  - supports multiple executions
  - each instance has it's own recorder
  - Editors opened from the script list, control the actual script in the list, and viceversa.
  - Stopping scripts from hotkey, api, etc stop also editor scripts ( as they are now just regular ScriptsEnhanced scripts )
  
 - all the other files: 
 minor changes to reflect the changes of API signatures, etc

### CODE REVIEW

File: Razor/RazorEnhanced/EnhancedScript.cs [hunk 1]
```diff
@@ -0,0 +1,873 @@
...
+            //if (IsRunning || !IsUnstarted)
+            if (IsRunning) return;
+
+            try
+            {
+                //EventManager.Instance.Unsubscribe(m_Thread); 
+                m_Thread.Start();
+                while (!m_Thread.IsAlive)
+                {
+                }
```
Comment (credzba) [inline Razor/RazorEnhanced/EnhancedScript.cs]: Ought to have at least a little pause in here

File: Razor/Network/Handlers.cs [hunk 2]
```diff
@@ -425,15 +417,13 @@ private static void ClientTextCommand(PacketReader p, PacketHandlerEventArgs arg
                     }
                 case 0xF4: // Invoke Virtue
                     {
-                        if (RazorEnhanced.ScriptRecorder.OnRecord)
+                        try
```
Comment (credzba) [inline Razor/Network/Handlers.cs:420]: I wonder why this has a try catch

File: Razor/RazorEnhanced/UI/EnhancedScriptEditor.cs [hunk 3]
```diff
@@ -45,40 +38,47 @@ private enum Command
...
+        private static Command m_CurrentCommand = Command.None;
+        private static readonly AutoResetEvent m_WaitDebug = new AutoResetEvent(false);
+
+        
+        
 
-        private string m_Filename = String.Empty;
-        private string m_Filetype = String.Empty;
-        public  static string GetFiletype() {  return m_EnhancedScriptEditor.m_Filetype; }
+        //private string m_Filename = String.Empty;
```
Comment (credzba) [inline Razor/RazorEnhanced/UI/EnhancedScriptEditor.cs:62]: delete comments. We can always get them from git if critical

File: Razor/RazorEnhanced/UI/EnhancedScriptEditor.cs [hunk 4]
```diff
@@ -1061,93 +979,118 @@ private void Save()
...
-                m_Filename = Path.GetFileName(save.FileName);
-                SavaData();
-                ReloadAfterSave();
+                var fullpath = save.FileName;
+                if (m_Script.Editor)
+                {
+                    EnhancedScript.RemoveScript(m_Script);
+                }
+                //Dalamar:
+                //TODO: add YES/NO dialog for overwrite check File.Exists(fullpath)
```
Comment (credzba) [inline Razor/RazorEnhanced/UI/EnhancedScriptEditor.cs:1017]: Would be nice

**Review by credzba — APPROVED** [review]
I didnt see anything obviously wrong. I had a few comments but they are just comment, not critical

File: Razor/Network/Handlers.cs (hunk 2 above)
Comment (cesare-montresor) [reply Razor/Network/Handlers.cs:420]: dunno, probably for the:
command.Split(' ')[0]

File: Razor/RazorEnhanced/UI/EnhancedScriptEditor.cs (hunk 4 above)
Comment (cesare-montresor) [reply Razor/RazorEnhanced/UI/EnhancedScriptEditor.cs:1017]: it's actaully aready included in the save dialog anytime you try to verride a file

File: Razor/RazorEnhanced/UI/EnhancedScriptEditor.cs (hunk 3 above)
Comment (cesare-montresor) [reply Razor/RazorEnhanced/UI/EnhancedScriptEditor.cs:62]: yes, I left them there for comparison, I'll do a global cleanup of what is left.

File: Razor/RazorEnhanced/EnhancedScript.cs (hunk 1 above)
Comment (cesare-montresor) [reply Razor/RazorEnhanced/EnhancedScript.cs]: I'll add a Misc.Pause(1)

**Pull request merged**